### ISSUE TITLE
Unless specified otherwise, keras.Input defaults to 'float32', regardless of the current 'floatx' setting

### ISSUE THREAD

**Issue by pass-lin** [keras-team/keras#19332]
This occurs because, for example, when inputting 2909, keras.Input will convert it to bf16 regardless of whether keras.mixed_precision.set_global_policy("mixed_bfloat16") or keras.backend.set_floatx('bfloat16') is set, resulting in a value of 2912 (2908 under fp16). Clearly, this can lead to significant errors in models, especially when using pre-trained language models. The solution is to use the following code:

def Input(
    shape=None,
    batch_size=None,
    dtype=None,
    sparse=None,
    batch_shape=None,
    name=None,
    tensor=None,
):  
      if dtype==None:  
          dtype='float32'  
      return keras.Input(  
          shape=shape,  
          batch_size=batch_size,  
          dtype=dtype,  
          sparse=sparse,  
          batch_shape=batch_shape,  
          name=name,  
          tensor=tensor,  
      )
  

Similarly, default target inputs may also encounter errors due to such conversions. Therefore, it is desired that input and target inputs default to fp32 under all circumstances unless explicitly specified.

**Comment by SuryanarayanaY** [comment]
Hi @pass-lin ,

The purpose of `keras.config.set_floatx` is to change the default dtype to higher or low precisions based on individual requirements which otherwise is float32 by default. With `mixed_precision` policy, layers use float16 computations and float32 variables for improving performance.

User has choice to decide whether to change default dtype or not.If you want to keep float32, no need to use no need to do any additional setting of `keras.config.set_floatx` or `mixed_precision` policy.

If user chooses low precision dtype then it is expected to lose some precision but might not be that much.If the input is not normalized it might have a little influence on precision. Thanks!

**Comment by pass-lin** [comment]
> @pass-lin
> 
> 嗨@pass-lin，
> 
> `keras.config.set_floatx``mixed_precision`
> 
> 使用`keras.config.set_floatx`的目的是根据个人需求将默认dtype更改为更高或更低精度，否则默认为float32。使用`mixed_precision`策略，图层使用float16计算和float32变量来提高性能。
> 
> `keras.config.set_floatx``mixed_precision`
> 
> 用户可以选择是否更改默认dtype。如果你想保留float32，不需要使用不需要做任何额外的设置`keras.config.set_floatx`或`mixed_precision`策略。
> 
> If user chooses low precision dtype then it is expected to lose some precision but might not be that much.If the input is not normalized it might have a little influence on precision. Thanks!
> 
> 如果用户选择低精度dtype，那么预计会损失一些精度，但可能不会那么多。如果输入未归一化，它可能会对精度产生一点影响。谢谢！
However, this will lead to errors. For example, if you input tf.tensor([2909],dtype=tf.int32), after passing through keras<KerasTensor shape=(None, None), dtype=bfloat16, sparse=None, name=keras_tensor>, it will become <tf.Tensor: shape=(1,), dtype=bfloat16, numpy=array([2912], dtype=bfloat16)>.
If the user wishes to modify the precision using keras.config.set_floatx without changing the code, errors will occur.

**Comment by pass-lin** [comment]
> Hi @pass-lin ,
> 
> The purpose of `keras.config.set_floatx` is to change the default dtype to higher or low precisions based on individual requirements which otherwise is float32 by default. With `mixed_precision` policy, layers use float16 computations and float32 variables for improving performance.
> 
> User has choice to decide whether to change default dtype or not.If you want to keep float32, no need to use no need to do any additional setting of `keras.config.set_floatx` or `mixed_precision` policy.
> 
> If user chooses low precision dtype then it is expected to lose some precision but might not be that much.If the input is not normalized it might have a little influence on precision. Thanks!

However, this will lead to errors. For example, if you input tf.tensor([2909],dtype=tf.int32), after passing through keras<KerasTensor shape=(None, None), dtype=bfloat16, sparse=None, name=keras_tensor>, it will become <tf.Tensor: shape=(1,), dtype=bfloat16, numpy=array([2912], dtype=bfloat16)>.
If the user wishes to modify the precision using keras.config.set_floatx without changing the code, errors will occur.

**Comment by SuryanarayanaY** [comment]
Hi @pass-lin ,
Could you submit a minimal code repro for this?

**Comment by pass-lin** [comment]
> Hi @pass-lin , Could you submit a minimal code repro for this?

this is a simple error code
`import os
os.environ["KERAS_BACKEND"] = "torch"
import keras
keras.backend.set_floatx('bfloat16')
x = keras.layers.Input([None],dtype='float32')
def temp(a):
    print(a,a.dtype)
    return a
z = keras.layers.Lambda(temp,dtype='float32')(x)
y = Embedding(2910,32)(x)
model=keras.Model(x,z)
model.summary()
import numpy as np
a = np.ones([1,3])*2908
model.predict(a)`

we can found [2909,2909,2909] has became tensor([[2912., 2912., 2912.]], device='cuda:0', dtype=torch.bfloat16)
if not fit this bug,keras.backend.set_floatx('bfloat16') and keras.config.set_dtype_policy("mixed_bfloat16") will be error.

**Comment by fchollet** [comment]
@pass-lin if you want to feed an `Embedding` layer, you need to pass `dtype="int32"` to your `Input`. Even float32 is incorrect in this case. This is not a bug -- you just need to pass to your `Input` the `dtype` that corresponds to the kind of data you expect.
